### ISSUE TITLE
Future work suggestions

### ISSUE THREAD

**Issue by IanQS** [NarainKrishnamurthy/BeatGAN2.0#1]
**Folder Structure**

Just personal opinions

- [ ] move some functions into a utils.py file

- [ ] store non-essential functions into folders. E.g have a folder called 'models' and within that have two separate files called 'discriminator.py' and 'generator.py' which store the Discriminator and Generator classes

- [ ] rename drums folder to 'old_dataset' or create a top level 'dataset' folder and move 'drums' into a V1.0 

**Analysis Suggestions**

I'd love to see more about WHY you chose that similarity metric of sum [(A-B)^2 / A^2]. What happens if you dig through sklearn and scipys signal analysis modules? 

My signal analysis is weak so these suggestions might not make sense (and it's admittedly more data science than ML) but what do similar signals sound like under metric X?

- [ ] Dynamic time warping -> 

- [ ] Mutual information

- [ ] etc

**Writeup suggestions**

- Your writeup is great! Perhaps part 2 belongs in a medium post but I'd love to hear about the process you went through - how was debugging it? Did you run into any large issues and if so how'd you get around them? I know GANs are a pain in the ass to train - did you refer to anything while debugging that someone who has a cursory interest might find useful while they actually implement it on their own?

**Comment by akaniklaus** [comment]
@IanQS DTW would be useful only if they have different length. But I agree that other similarity metrics can maybe lead to better results. Some that I can think of is cross correlation and cosine distance. I believe that there are also some Fast Fourier Transform-based similarity methods for audio signals. I would actually be interested in generating beats which would progressively follow up each other. Or for example generating intermediate beats when two beats are given (I guess that would be a path finding problem when the similarity metric is given?). As far as I checked the model predictions, the biggest problem seems to be that most of generated signals are similar or identical to training samples as @NarainKrishnamurthy mentioned in the readme document. One can generate several beats with random inputs and select most distinctive N via k-Medoids algorithm via the similarity metric defined.

**Comment by akaniklaus** [comment]
@NarainKrishnamurthy I am new to GANs. Have you utilized minibatch discrimination? I see from the code that you use random weighted averaged samples for discrimination, is this for that purpose?

**Comment by NarainKrishnamurthy** [comment]
@akaniklaus The weighted average is used to implement the Gradient Penalty part of the WGAN-GP loss. Minibatch discrimination is not used. That technique is generally used to ensure diversity in the Generator's distribution and diversity wasn't really a problem (relative to the training data). Really, the best option is likely more data which, in practice, means more meaningful data augmentation and better generation of training examples from individual beats (e.g. one beat maps to 4 samples).
